File: src/python/ch01/1.4/l1/fixed_point.py
from dataclasses import dataclass

# Money as fixed-point cents (int64 policy). Safe up to about ±9e16 cents.
def parse_dollars_to_cents(s: str) -> int:
    """
    Parse strings like '123', '123.4', '123.45', '-0.07' to cents (int).
    Truncates extra fractional digits beyond 2 (policy: banker-friendly).
    """
    s = s.strip()
    sign = -1 if s.startswith('-') else 1
    if s[0] in '+-':
        s = s[1:]
    if '.' in s:
        whole, frac = s.split('.', 1)
    else:
        whole, frac = s, ''
    if not whole:
        whole = '0'
    if not whole.isdigit() or (frac and not frac.isdigit()):
        raise ValueError(f"bad money literal: {s}")
    frac = (frac + '00')[:2]  # pad/truncate to 2 decimals
    return sign * (int(whole) * 100 + int(frac))

def format_cents(c: int) -> str:
    sign = '-' if c < 0 else ''
    c = abs(c)
    return f"{sign}{c//100}.{c%100:02d}"

@dataclass
class Ledger:
    """In-memory ledger with int cents; raises on overdraft by default."""
    balance_cents: int = 0

    def deposit(self, cents: int) -> None:
        # No overflow in Python, but you can bound-check here if porting.
        self.balance_cents += cents

    def withdraw(self, cents: int) -> None:
        if cents > self.balance_cents:
            raise ValueError("insufficient funds")
        self.balance_cents -= cents

    def transfer_to(self, other: "Ledger", cents: int) -> None:
        self.withdraw(cents)
        other.deposit(cents)

File: src/go/ch01/1.4/l1/fixed_point.go
package main

import (
	"errors"
	"fmt"
	"strings"
	"unicode"
)

// ParseDollarsToCents parses "123", "123.4", "123.45", "-0.07" into cents (int64).
// Extra fractional digits are truncated by policy.
func ParseDollarsToCents(s string) (int64, error) {
	s = strings.TrimSpace(s)
	if s == "" {
		return 0, errors.New("empty money literal")
	}
	sign := int64(1)
	if s[0] == '+' || s[0] == '-' {
		if s[0] == '-' {
			sign = -1
		}
		s = s[1:]
		if s == "" {
			return 0, errors.New("bad money literal")
		}
	}

	whole := int64(0)
	i := 0
	for i < len(s) && unicode.IsDigit(rune(s[i])) {
		d := int64(s[i] - '0')
		// whole = whole*10 + d (checked with 64-bit bounds)
		if z, ok := MulI64Checked(whole, 10); ok {
			if z2, ok2 := AddI64Checked(z, d); ok2 {
				whole = z2
			} else {
				return 0, errors.New("overflow parsing whole")
			}
		} else {
			return 0, errors.New("overflow parsing whole")
		}
		i++
	}

	d1, d2 := int64(0), int64(0)
	if i < len(s) && s[i] == '.' {
		i++
		if i < len(s) && unicode.IsDigit(rune(s[i])) {
			d1 = int64(s[i] - '0')
			i++
		}
		if i < len(s) && unicode.IsDigit(rune(s[i])) {
			d2 = int64(s[i] - '0')
			i++
		}
		// ignore extra fractional digits
		for i < len(s) && unicode.IsDigit(rune(s[i])) {
			i++
		}
	}

	if strings.TrimSpace(s[i:]) != "" {
		return 0, errors.New("trailing junk in money literal")
	}

	cents := int64(0)
	if z, ok := MulI64Checked(whole, 100); ok {
		if z2, ok2 := AddI64Checked(z, d1*10+d2); ok2 {
			cents = z2
		} else {
			return 0, errors.New("overflow combining cents")
		}
	} else {
		return 0, errors.New("overflow combining cents")
	}

	return sign * cents, nil
}

// FormatCents prints cents to "[-]W.DD".
func FormatCents(c int64) string {
	sign := ""
	if c < 0 {
		sign = "-"
		c = -c
	}
	return fmt.Sprintf("%s%d.%02d", sign, c/100, c%100)
}

type Ledger struct {
	BalanceCents int64
}

func (l *Ledger) Deposit(cents int64) error {
	if z, ok := AddI64Checked(l.BalanceCents, cents); ok {
		l.BalanceCents = z
		return nil
	}
	return errors.New("overflow on deposit")
}

func (l *Ledger) Withdraw(cents int64) error {
	if cents > l.BalanceCents {
		return errors.New("insufficient funds")
	}
	l.BalanceCents -= cents
	return nil
}

func (l *Ledger) TransferTo(other *Ledger, cents int64) error {
	if err := l.Withdraw(cents); err != nil {
		return err
	}
	if err := other.Deposit(cents); err != nil {
		// rollback on deposit failure
		l.BalanceCents += cents
		return err
	}
	return nil
}
